zipper/protocols/irondb/irondb_group.go

```diff
@@ -0,0 +1,708 @@
+	var result []string
+	var target string
+	var tag_category string
+
+	// decoding query
+	queryDecoded, _ := url.QueryUnescape(query)
+	querySplit := strings.Split(queryDecoded, "&")
+	for _, qvRaw := range querySplit {
+		idx := strings.Index(qvRaw, "=")
+		//no parameters passed
+		if idx < 1 {
+			continue
+		}
+		k := qvRaw[:idx]
+		v := qvRaw[idx+1:]
+		if v2, ok := params[qvRaw[:idx]]; !ok {
+			params[k] = []string{v}
+		} else {
+			v2 = append(v2, v)
+			params[k] = v2
+		}
+	}
+	logger.Debug("doTagQuery",
+		zap.Any("query", queryDecoded),
+		zap.Bool("isTagName", isTagName),
+		zap.Any("params", params),
+		zap.Int64("limit", limit),
+	)
+
+	// default target - all Graphite metrics
+	target = `__name=~.*`
+	// but use joined expr value if present
+	if len(params["expr"]) > 0 {
+		target = strings.Join(params["expr"], ",")
+	}
+	if isTagName {
+		logger = logger.With(zap.String("type", "tagName"))
+	} else {
+		logger = logger.With(zap.String("type", "tagValues"))
+		// get tag category (if present and we're looking for values instead categories)
+		if tag, ok := params["tag"]; ok {
+			tag_category = tag[0]
+		} else {
+			return []string{}, types.ErrNoTagSpecified
+		}
+	}
+
+	logger.Debug("sending GraphiteFindTags request to irondb",
+		zap.Int64("accountID", c.accountID),
+		zap.String("prefix", c.graphite_prefix),
+		zap.String("target", target),
+	)
+	tag_result, err := c.client.GraphiteFindTags(c.accountID, c.graphite_prefix, target, nil)
+	if err != nil {
+		return []string{}, merry.New("request returned an error").WithValue("error", err)
+	}
+	logger.Debug("got GraphiteFindTags result from irondb",
+		zap.String("target", target),
+		zap.Any("tag_result", tag_result),
+	)
+	for _, metric := range tag_result {
+		// if metric name contain tags
+		if strings.Contains(metric.Name, ";") {
+			namex := strings.Split(metric.Name, ";")
+			for i := 1; i < len(namex); i++ {
+				tag := strings.SplitN(namex[i], "=", 2)
+				// tag[0] is tag category and tag[1] is tag value
+				if isTagName {
+					// if filtering by tag prefix
+					if v, ok := params["tagPrefix"]; ok {
+						// and prefix match
+						if strings.HasPrefix(tag[0], v[0]) {
+							// append tag name to result
+							result = append(result, tag[0])
+						}
+					} else {
+						// if not filtering by tag prefix - append all tags
+						result = append(result, tag[0])
+					}
+				} else {
+					// if filtering by value prefix
+					if v, ok := params["valuePrefix"]; ok {
+						// and prefix match
+						if strings.HasPrefix(tag[1], v[0]) {
+							// append tag value to result
+							result = append(result, tag[1])
+						}
+					} else {
+						// if not filtering by value prefix
+						// append only values belong to requested tag category to result
+						if tag[0] == tag_category {
+							result = append(result, tag[1])
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// removing duplicates from result list
+	seen := make(map[string]struct{}, len(result))
```

Comment:
It's fine to have it like that, but it's a potential candidate to be optimized as you can combine populating seen with the populating results. But it will make code more complicated.

Reply:
I would prefer to keep it like this, it's complicated enough. If I found way how to optimize it - I'll include dedup too